contracts/staking/StakingManager.sol

```diff
@@ -0,0 +1,447 @@
+            );
+            uptimeSeconds = uptime;
+        }
+        validator.uptimeSeconds = uptimeSeconds;
+
+        // Save the validator updates.
+        // TODO: Optimize storage writes here (probably don't need to write the whole value).
+        $._validationPeriods[validationID] = validator;
+
+        // Submit the message to the Warp precompile.
+        bytes memory setValidatorWeightPayload = StakingMessages.packSetSubnetValidatorWeightMessage(
+            validationID, validator.messageNonce, 0
+        );
+        bytes32 messageID = WARP_MESSENGER.sendWarpMessage(setValidatorWeightPayload);
+
+        // Emit the event to signal the start of the validator removal process.
+        emit ValidatorRemovalInitialized(
+            validationID, messageID, validator.weight, block.timestamp, uptimeSeconds
+        );
+    }
+
+    /**
+     * @notice Resubmits a validator end message to be sent to P-Chain to the Warp precompile.
+     * Only necessary if the original message can't be delivered due to validator churn.
+     */
+    // solhint-disable-next-line no-empty-blocks
+    function resendEndValidatorMessage(bytes32 validationID) external {
+        StakingManagerStorage storage $ = _getStakingManagerStorage();
+        Validator memory validator = $._validationPeriods[validationID];
+
+        require(
+            validator.status == ValidatorStatus.PendingRemoved,
+            "StakingManager: Validator not pending removal"
+        );
+
+        bytes memory setValidatorWeightPayload = StakingMessages.packSetSubnetValidatorWeightMessage(
+            validationID, validator.messageNonce, 0
+        );
+        WARP_MESSENGER.sendWarpMessage(setValidatorWeightPayload);
+    }
+
+    /**
+     * @notice Completes the process of ending a validation period by receiving an acknowledgement from the P-Chain
+     * that the validation ID is not active and will never be active in the future. Returns the the stake associated
+     * with the validation. Note that this function can be used for successful validation periods that have been explicitly
+     * ended by calling {initializeEndValidation} or for validation periods that never began on the P-Chain due to the
+     * {registrationExpiry} being reached.
+     * @param setWeightMessageType Whether or not the message type is a SetValidatorWeight message, or a
+     * SubnetValidatorRegistration message (with valid set to false). Both message types are valid for ending
+     * a validation period.
+     */
+    function completeEndValidation(uint32 messageIndex, bool setWeightMessageType) external {
+        StakingManagerStorage storage $ = _getStakingManagerStorage();
+
+        // Get the Warp message.
+        (WarpMessage memory warpMessage, bool valid) =
+            WARP_MESSENGER.getVerifiedWarpMessage(messageIndex);
+        require(valid, "StakingManager: Invalid warp message");
+
+        bytes32 validationID;
+        if (setWeightMessageType) {
```

Comment:
did we confirm that both SetWeightTx=0 and InvalidMsg can be signed for this one? I wonder if it makes a better UX to expect a single msg type to be signed/relayed.